src/compile.rs

```diff
@@ -459,6 +459,197 @@ pub fn validate_rules(statements: &BTreeSet<ValidatedStatement>) -> Result<(), C
+                                continue 'outer;
+                            }
+                            if let Some(ti) = types.get(arg.as_ref()) {
+                                if ti.is_associated_resource(types) {
+                                    num_changed += 1;
+                                    func.is_castable = false;
+                                    continue 'outer;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    num_changed
+}
+
+pub fn initialize_terminated<'a>(
+    functions: &'a FunctionMap<'a>,
+) -> (Vec<String>, Vec<FunctionInfo<'a>>) {
+    let mut term_ret_vec: Vec<String> = Vec::new();
+    let mut nonterm_ret_vec: Vec<FunctionInfo> = Vec::new();
+
+    for func in functions.values() {
+        let mut is_term = true;
+
+        for call in func.original_body {
+            if let Statement::Call(call) = call {
+                match call.check_builtin() {
+                    Some(_) => {
+                        continue;
+                    }
+                    None => {
+                        is_term = false;
+                        break;
+                    }
+                }
+            }
+        }
+        if is_term {
+            term_ret_vec.push(func.get_cil_name().clone());
+        } else {
+            nonterm_ret_vec.push(func.clone());
+        }
+    }
+
+    (term_ret_vec, nonterm_ret_vec)
+}
+
+pub fn search_for_recursion(
+    terminated_list: &mut Vec<String>,
+    functions: &mut Vec<FunctionInfo>,
+) -> Result<(), CascadeErrors> {
+    let mut removed: u64 = 1;
+    while removed > 0 {
+        removed = 0;
+        for func in functions.clone().iter_mut() {
+            let mut is_term = true;
+            for call in func.original_body {
+                if let Statement::Call(call) = call {
+                    let mut call_cil_name = call.get_cil_name();
+                    // If we are calling something with this it must be in the same class
+                    // so hand place the class name
+                    if call_cil_name.contains("this-") {
+                        if let Some(class) = func.class {
+                            call_cil_name =
+                                call_cil_name.replace("this-", &(class.name.to_string() + "-"))
+                        } else {
+                            return Err(CascadeErrors::from(ErrorItem::make_compile_or_internal_error(
+                                "Could not determine class for 'this.' function call",
+                                Some(func.declaration_file),
+                                call.get_name_range(),
+                                "Perhaps you meant to place the function in a resource or domain?")));
+                        }
+                    }
+
+                    if terminated_list.contains(&call_cil_name) {
```

Comment:
Why not reverse the logic and set is_term=false break if terminated_list does *not* contain the call?  Then we'll short circuit and not keep checking the remaining calls after the first non-terminated one?

Reply:
I don't think the new update matches the suggestion.  Maybe I'm missing something, but it looks like we're breaking and assume we terminate if we find a single terminating function.  The suggestion above is to assume we are terminating, then if we see a single non-terminating function, set is_term = false and break immediately, which doesn't look like what the code does now.

Reply:
Isn't that the case that if we do find a single terminating thing we are ourselves terminating?  If we find a non-terminating thing there still may be other terminating things to find.  Or do I have my logic just 100% backwards on that?

Reply:
I'm pretty sure you're 100% backwards.  If a function calls any non-terminating functions, then it's non-terminating.

```
fn foo {
    allow(...); // terminating
    allow(...); // terminating
    // etc, more terminating functions
    bar(...); // non-terminating
}

fn bar {
    foo(...); non-terminating
}
```

foo/bar are in a recursive loop, so foo is non-terminating, regardless of the presence of other terminating functions.

Reply:
At a first read the logic as it stands now seems correct to me.